geopandas/_vectorized.py

```diff
@@ -0,0 +1,853 @@
+    else:
+        return pygeos.from_wkb(geom.wkb)
+
+
+def from_shapely(data):
+    """
+    Convert a list or array of shapely objects to an object-dtype numpy
+    array of validated geometry elements.
+
+    """
+    n = len(data)
+
+    out = []
+
+    for idx in range(n):
+        geom = data[idx]
+        if compat.USE_PYGEOS and isinstance(geom, pygeos.Geometry):
+            out.append(geom)
+        elif isinstance(geom, BaseGeometry):
+            if compat.USE_PYGEOS:
+                out.append(_shapely_to_pygeos(geom))
+            else:
+                out.append(geom)
+        elif hasattr(geom, "__geo_interface__"):
+            geom = shapely.geometry.asShape(geom)
+            if compat.USE_PYGEOS:
+                out.append(_shapely_to_pygeos(geom))
+            else:
+                out.append(geom)
+        elif _isna(geom):
+            out.append(None)
+        else:
+            raise TypeError("Input must be valid geometry objects: {0}".format(geom))
+
+    if compat.USE_PYGEOS:
+        return np.array(out, dtype=object)
+    else:
+        aout = np.empty(n, dtype=object)
+        aout[:] = out
+        return aout
+
+
+def to_shapely(data):
+    if compat.USE_PYGEOS:
+        out = np.empty(len(data), dtype=object)
+        out[:] = [_pygeos_to_shapely(geom) for geom in data]
+        return out
+    else:
+        return data
+
+
+def from_wkb(data):
+    """
+    Convert a list or array of WKB objects to a np.ndarray[geoms].
+    """
+    if compat.USE_PYGEOS:
+        return pygeos.from_wkb(data)
+
+    import shapely.wkb
+
+    n = len(data)
+
+    out = []
+
+    for idx in range(n):
+        geom = data[idx]
+        if geom is not None and len(geom):
+            geom = shapely.wkb.loads(geom)
+        else:
+            geom = None
+        out.append(geom)
+
+    aout = np.empty(n, dtype=object)
+    aout[:] = out
+    return aout
+
+
+def to_wkb(data):
+    if compat.USE_PYGEOS:
+        return pygeos.to_wkb(data)
+    else:
+        out = [geom.wkb if geom is not None else None for geom in data]
+        return np.array(out, dtype=object)
+
+
+def from_wkt(data):
+    """
+    Convert a list or array of WKT objects to a np.ndarray[geoms].
+    """
+    if compat.USE_PYGEOS:
+        return pygeos.from_wkt(data)
+
+    import shapely.wkt
+
+    n = len(data)
+
+    out = []
+
+    for idx in range(n):
+        geom = data[idx]
+        if geom is not None and len(geom):
+            if isinstance(geom, bytes):
+                geom = geom.decode("utf-8")
+            geom = shapely.wkt.loads(geom)
+        else:
+            geom = None
+        out.append(geom)
+
+    aout = np.empty(n, dtype=object)
+    aout[:] = out
+    return aout
+
+
+def to_wkt(data, **kwargs):
+    if compat.USE_PYGEOS:
+        return pygeos.to_wkt(data, **kwargs)
+    else:
+        out = [geom.wkt if geom is not None else None for geom in data]
+        return np.array(out, dtype=object)
```

Comment:
Same as above, inconsistent pattern in array creation.

Reply:
Here we are not creating geometries (but we have wkt strings), so can't have the problem of numpy expanding multi-geoms